drivers/i2s/i2s_nrfx.c

```diff
@@ -0,0 +1,927 @@
+	     (drv_data->tx.cfg.block_size != drv_data->rx.cfg.block_size))) {
+		LOG_ERR("TX and RX configurations are different");
+		return -EIO;
+	}
+
+	switch (cmd) {
+	case I2S_TRIGGER_START:
+		cmd_allowed = (drv_data->state == I2S_STATE_READY);
+		break;
+	case I2S_TRIGGER_STOP:
+	case I2S_TRIGGER_DRAIN:
+		cmd_allowed = (drv_data->state == I2S_STATE_RUNNING);
+		break;
+	case I2S_TRIGGER_DROP:
+		cmd_allowed = configured;
+		break;
+	case I2S_TRIGGER_PREPARE:
+		cmd_allowed = (drv_data->state == I2S_STATE_ERROR);
+		break;
+	default:
+		LOG_ERR("Invalid trigger: %d", cmd);
+		return -EINVAL;
+	}
+
+	if (!cmd_allowed) {
+		return -EIO;
+	}
+
+	/* For triggers applicable to the RUNNING state (i.e. STOP, DRAIN,
+	 * and DROP), ensure that the command is applied to the streams
+	 * that are currently active (this device cannot e.g. stop only TX
+	 * without stopping RX).
+	 */
+	if (drv_data->state == I2S_STATE_RUNNING &&
+	    drv_data->active_dir != dir) {
+		LOG_ERR("Inappropriate trigger (%d/%d), active stream(s): %d",
+			cmd, dir, drv_data->active_dir);
+		return -EINVAL;
+	}
+
+	switch (cmd) {
+	case I2S_TRIGGER_START:
+		drv_data->stop = false;
+		drv_data->discard_rx = false;
+		drv_data->active_dir = dir;
+		drv_data->next_tx_buffer_needed = false;
+		return trigger_start(dev);
+
+	case I2S_TRIGGER_STOP:
+		drv_data->state = I2S_STATE_STOPPING;
+		drv_data->stop = true;
+		return 0;
+
+	case I2S_TRIGGER_DRAIN:
+		drv_data->state = I2S_STATE_STOPPING;
+		/* If only RX is active, DRAIN is equivalent to STOP. */
+		drv_data->stop = (drv_data->active_dir == I2S_DIR_RX);
+		return 0;
+
+	case I2S_TRIGGER_DROP:
+		if (drv_data->state != I2S_STATE_READY) {
+			drv_data->discard_rx = true;
+			nrfx_i2s_stop();
+		}
+		purge_queue(dev, dir);
+		drv_data->state = I2S_STATE_READY;
+		return 0;
+
+	case I2S_TRIGGER_PREPARE:
+		purge_queue(dev, dir);
+		drv_data->state = I2S_STATE_READY;
+		return 0;
+
+	default:
+		LOG_ERR("Invalid trigger: %d", cmd);
+		return -EINVAL;
+	}
+}
+
+static void init_clock_manager(const struct device *dev)
+{
+	struct i2s_nrfx_drv_data *drv_data = dev->data;
+	clock_control_subsys_t subsys;
+
+#if NRF_CLOCK_HAS_HFCLKAUDIO
+	const struct i2s_nrfx_drv_cfg *drv_cfg = dev->config;
+
+	if (drv_cfg->clk_src == ACLK) {
+		subsys = CLOCK_CONTROL_NRF_SUBSYS_HFAUDIO;
+	} else
+#endif
+	{
+		subsys = CLOCK_CONTROL_NRF_SUBSYS_HF;
+	}
+
+	drv_data->clk_mgr = z_nrf_clock_control_get_onoff(subsys);
+	__ASSERT_NO_MSG(drv_data->clk_mgr != NULL);
+}
+
+static const struct i2s_driver_api i2s_nrf_drv_api = {
+	.configure = i2s_nrfx_configure,
+	.config_get = i2s_nrfx_config_get,
+	.read = i2s_nrfx_read,
+	.write = i2s_nrfx_write,
+	.trigger = i2s_nrfx_trigger,
+};
+
+#define I2S(idx) DT_NODELABEL(i2s##idx)
+
+#define I2S_PIN(idx, name)					\
+	COND_CODE_1(DT_NODE_HAS_PROP(I2S(idx), name##_pin),	\
+		    (DT_PROP(I2S(idx), name##_pin)),		\
+		    (NRFX_I2S_PIN_NOT_USED))
+#define I2S_CLK_SRC(idx)  DT_ENUM_IDX(I2S(idx), clock_source)
```

Comment:
What about using `DT_ENUM_UPPER_TOKEN()` instead, so that we don't need to keep enum order in sync between DT bindings and this driver?

Reply:
Good point, thanks! Applied.